File: Baekjoon Online Judge/Gold/BOJ 20040.py
import sys
input=sys.stdin.readline
sys.setrecursionlimit(1000000)#아니 왜 내거만 재귀깊이 걸리냐고
def getpa(papa, n):
	if(papa[n]==n): return n
	papa[n] = getpa(papa,papa[n])
	return papa[n]
N,M = map(int,input().split())
papa = list(range(N))
ans = 0
for i in range(M):
	src,dest=input().split()
	src = int(src)
	dest = int(dest)
	if(getpa(papa,src)==getpa(papa,dest)):
		ans=i+1
		break
	papa[getpa(papa,dest)] = papa[src]
print(ans)

File: baekjoon_online_judge/800_gold/4_950/BOJ 20040.c
int getpapa(int *papa, int n) {
	if (papa[n] == n) return n;
	return papa[n] = getpapa(papa, papa[n]);
}
int main() {
	int N, M;
	int i, src, dest, ans=0;
	int* papa;
	scanf("%d%d", &N, &M);
	papa = malloc(sizeof(int*)*N);

	for (i = 0; i < N; i++)
		papa[i] = i;

	for (i = 0; i < M; i++) {
		scanf("%d%d", &src, &dest);
		if (getpapa(papa, src) == getpapa(papa, dest)) {
			ans = i + 1;
			break;
		}
		papa[getpapa(papa,dest)] = papa[src];
	}
	printf("%d", ans);
}
